apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/Interval.cs

```diff
@@ -0,0 +1,314 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.CodeDom;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Google.Cloud.Spanner.V1;
+
+/// <summary>
+/// Representation of a Time Interval.
+/// </summary>
+public class Interval
+{
+    private const int Uninitialized = 0;
+    private const int ParsingDate = 1;
+    private const int ParsingTime = 2;
+    internal const long NanosecondsInASecond = 1000000000L;
+    internal const long NanosecondsInAMinute = NanosecondsInASecond * 60;
+    internal const long NanosecondsInAnHour = NanosecondsInAMinute * 60;
+    internal const long NanosecondsInADay = NanosecondsInAnHour * 24;
+    internal const long NanosecondsInAMonth = NanosecondsInADay * 30;
+    internal const long NanosecondsInAYear = NanosecondsInAMonth * 12;
+    internal const long NanosecondsInAMillisecond = 1000000L;
+    internal const long NanosecondsInAMicrosecond = 1000L;
+    internal const int MaxMonths = 120000;
+    internal const int MinMonths = -MaxMonths;
+    internal const int MaxDays = 3660000;
+    internal const int MinDays = -MaxDays;
+
+    /// <summary>
+    /// Month portion of this interval.
+    /// </summary>
+    public int Months { get; }
+
+    /// <summary>
+    /// Day portion of this interval.
+    /// </summary>
+    public int Days { get; }
+
+    /// <summary>
+    /// Nanosecond portion of this interval.
+    /// </summary>
+    public BigInteger Nanoseconds { get; }
+
+    /// <summary>
+    /// Representation of a Time Interval.
+    /// </summary>
+    private Interval(int months, int days, BigInteger nanoseconds)
+    {
+        if (months > MaxMonths || months < MinMonths)
+        {
+            throw new ArgumentException($"The interval type supports a range from {MinMonths} to {MaxMonths} months");
+        }
+        Months = months;
+
+        if (days > MaxDays || days < MinDays)
+        {
+            throw new ArgumentException($"The interval type supports a range from {MinDays} to {MaxDays} days");
+        }
+        Days = days;
+
+        BigInteger MaxNanoseconds = BigInteger.Parse("316224000000000000000");
+
+        if (nanoseconds > MaxNanoseconds || nanoseconds < -MaxNanoseconds)
+        {
+            throw new ArgumentException($"The interval type supports a range from -{MaxNanoseconds} to {MaxNanoseconds} nanoseonds");
+        }
+        Nanoseconds = nanoseconds;
+    }
+
+    /// <summary>
+    /// Constructs an interval based on an ISO8601 duration string.
+    /// This should be formatted in the following manner:
+    /// P[n]Y[n]M[n]DT[n]H[n]M[n[.fraction]]S
+    /// where [n] is an int.
+    /// </summary>
+    public static Interval Parse(string intervalString)
```

Comment:
@Hectorhammett  see how this looks like. No Regex and I think it's pretty clear.

It's a console app so it needs tweaking to be used here, and I stopped at the point of extracting each of the parts, that is I didn't combined the parts in the 3 fields months, days and nanos, but that should be easy enough.

Code here: https://gist.github.com/amanda-tarafa/af0f21b45c7a461fcefd11a3e6ca8136

Reply:
Yes, I like this approach.

My NodaTime code doing similar work is here, but Amanda's is cleaner, I think.
https://github.com/nodatime/nodatime/blob/main/src/NodaTime/Text/PeriodPattern.cs#L215

(It's possible that I don't detect a period ending in 'T' though - I need to check that at some point.)

Reply:
I like this one over the regex as well. Let me take a jab at it and run the tests and will update the code accordingly :)